src/index.js

```diff
@@ -21,11 +25,13 @@ export default function createCachedSelector(...funcs) {
       selectorCreator = options.selectorCreator || createSelector;
     }
 
+    const isCacheKeyValid = cache.isCacheKeyValid || defaultCacheKeyValidator;
+
+    // Application receives this function
     const selector = function(...args) {
-      // Application receives this function
       const cacheKey = resolver(...args);
 
-      if (typeof cacheKey === 'string' || typeof cacheKey === 'number') {
+      if (isCacheKeyValid(cacheKey)) {
```

Comment:
I would very much like giving a warning if the key is invalid. Generally, it means that something is wrong in my key implementation. In fact, I would consider `undefined` as a suspicious key in all cases (even though Maps support it as a valid key).
```
console.warn(`[re-reselect] Invalid cache key '${cacheKey}' has been returned by resolver function.`);
```
Giving it a second thought, returning `undefined` as the result as a selector might lead to bugs which are very difficult to trace back in this case. Should we maybe throw an exception in this case? I tend to be rather restrictive, just an idea.

Reply:
I'd start with a `console.warn`, good point.
Wouldn't be an exception a bit too punitive since we are just deriving data?

Reply:
My reasoning for throwing an exception is this: according to current behaviour a cached selector will return `undefined` if the resolved key is invalid. This value will then be used in the user code and likely to throw a `TypeError` exception when accessed:
```javascript
const user = getUserDetailsById(state, userId); // cached with re-reselect by userId
return {
    email: user.email // TypeError thrown
};
```
So the earlier we indicate to the source of the problem, the easier it will be to avoid long debugging trying to find the source of the `undefined`.
To sum up, I would say that warning would be enough if we still returned a correct result and didn't break the user code apart from caching.